### ISSUE TITLE
Update to use dry-result_matcher

### ISSUE THREAD

**Issue by solnic** [dry-rb/dry-transaction#31]
Once result_matcher has support for matching on specific error type we can update this lib to use it and remove some code here

**Comment by mrbongiolo** [comment]
It should be something like this?

``` ruby
Customer::Create.call(params) do |m|
  m.success { |output| return a 200 }
  m.failure(Error::EntityNotFound) { |error| return a 404 }
  m.failure(Error::UnprocessableEntity) { |error| return a 422 }
  m.failoure { |error| return a 500 }
end

# or maybe

Customer::Create.call(params) do |m|
  m.success { |output| return a 200 }
  m.failure(:authorization) { |error| return a 403 }
  m.failure(:validation) { |error| return a 422 }
  m.failure { |error return a 500 }
end
```

**Comment by timriley** [comment]
I think, to start with, we would simply want to maintain dry-transaction's current API and behaviour, and establish a way in dry-result_matcher to support that.

Then, that'll probably open up some other possibilities for us.

**Comment by mrbongiolo** [comment]
The usage of the `step_name` is ok to start with, but as I was trying to use it, it felt a bit "strange", having to know the names of the step wasn't right. And sometimes I had a single transaction that could "validate" things in 2 or 3 different steps, that's why I ended up sending a `Left(Exception)` and then just processing it on a default failure block depending on the type.
How are you thinking on the API to define the different types of failures?
Should the step hold it? Ex: `step :something, failure: :authorization`
Should the Monad know about it?

**Comment by timriley** [comment]
This will depend on https://github.com/dry-rb/dry-result_matcher/issues/3 to be implemented (and be implemented flexibly enough for dry-transaction to use it).

Just noting this so we can continue the discussion in one place, over in this other issue :)

**Comment by timriley** [comment]
Resolved in #33
